tests/devices/test_caching.py

```diff
@@ -260,3 +260,38 @@ def qfunc(x, y):
         assert calls2 == 5
         assert calls3 == 10
         assert g is not None
+
+    def test_different_return_type(self):
+        """Test that same circuit with different return type, returns different results"""
+        nwires = 2
+        wires = range(nwires)
+
+        hamiltonian = np.array(
+            [
+                [-2.5623 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.0 - 0.1234j],
+                [0.0 + 0.0j, -2.5623 + 0.0j, 0.0 + 0.1234j, 0.0 + 0.0j],
+                [0.0 + 0.0j, 0.0 - 0.1234j, -2.5623 + 0.0j, 0.0 + 0.0j],
+                [0.0 + 0.1234j, 0.0 + 0.0j, 0.0 + 0.0j, -2.5623 + 0.0j],
+            ]
+        )
+
+        np.random.seed(172)
+        params = np.random.randn(2, nwires)
+        devs = [
+            qml.device("default.qubit", wires=nwires),
+            qml.device("default.qubit", cache=1, wires=nwires),
+        ]
+
+        for dev in devs:
```

Comment:
Maybe it's good to using `@pytest.mark.parametrize` to run this test twice for each device.